internal/status/httproutestatus.go

```diff
@@ -0,0 +1,143 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package status
+
+import (
+	"fmt"
+	"time"
+
+	contour_api_v1 "github.com/projectcontour/contour/apis/projectcontour/v1"
+	v1 "github.com/projectcontour/contour/apis/projectcontour/v1"
+	"github.com/projectcontour/contour/internal/k8s"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	gatewayapi_v1alpha1 "sigs.k8s.io/gateway-api/apis/v1alpha1"
+)
+
+type HTTPRouteUpdate struct {
+	Fullname           types.NamespacedName
+	Conditions         []metav1.Condition
+	ExistingConditions []metav1.Condition
+	GatewayRef         types.NamespacedName
+}
+
+// ConditionFor returns a v1.Condition for a given ConditionType.
+func (pu *HTTPRouteUpdate) ConditionFor(cond ConditionType, reason string, message string) metav1.Condition {
+	newDc := v1.Condition{}
+	newDc.Reason = reason
+	newDc.Status = contour_api_v1.ConditionTrue
+	newDc.Type = string(cond)
+
+	switch cond {
+	case AdmittedConditionValid:
+		newDc.Type = "Admitted"
+	case AdmittedConditionInvalid:
+		newDc.Status = contour_api_v1.ConditionFalse
+		newDc.Type = "Admitted"
+	}
+	newDc.Message = message
+	newDc.LastTransitionTime = metav1.NewTime(time.Now())
+	pu.Conditions = append(pu.Conditions, newDc)
+	return newDc
+}
+
+// HTTPRouteAccessor returns a HTTPRouteUpdate that allows a client to build up a list of
+// v1.Conditions as well as a function to commit the change back to the cache when everything
+// is done. The commit function pattern is used so that the HTTPRouteUpdate does not need
+// to know anything the cache internals.
+func (c *Cache) HTTPRouteAccessor(route *gatewayapi_v1alpha1.HTTPRoute) (*HTTPRouteUpdate, func()) {
+	pu := &HTTPRouteUpdate{
+		Fullname:           k8s.NamespacedNameOf(route),
+		Conditions:         []metav1.Condition{},
+		ExistingConditions: c.getGatewayConditions(route.Status.Gateways),
+		GatewayRef:         c.gatewayRef,
+	}
+
+	return pu, func() {
+		c.commitHTTPRoute(pu)
+	}
+}
+
+func (c *Cache) commitHTTPRoute(pu *HTTPRouteUpdate) {
+	if len(pu.Conditions) == 0 {
+		return
+	}
+	c.httpRouteUpdates[pu.Fullname] = pu
+}
+
+func (pu *HTTPRouteUpdate) Mutate(obj interface{}) interface{} {
```

Comment:
tiniest of nits: seems like `pu` as a variable name is used in various places and might have been copied from the proxy update type, not a big deal to leave in

Reply:
👍